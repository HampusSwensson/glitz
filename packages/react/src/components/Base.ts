import { StyleArray, StyleOrStyleArray } from '@glitz/type';
import * as React from 'react';
import { Consumer, Context } from './context';

export const APPLY_CLASS_NAME_IDENTIFIER = '__b$glitz';

export type CSSProp = {
  css?: StyleOrStyleArray;
};

export type StyledProps = {
  apply: () => string | undefined;
  compose: (style?: StyleOrStyleArray) => StyleOrStyleArray;
};

export type StyledElementProps = {
  className?: string;
};

export type ExternalProps<TProps> = Pick<TProps, Exclude<keyof TProps, keyof StyledProps>> & CSSProp;

export interface StyledComponent<TProps> extends React.ComponentClass<ExternalProps<TProps> & CSSProp> {
  compose: (style?: StyleArray) => StyledComponent<TProps>;
}

export type InnerRefProp<TInstance> = {
  innerRef?: React.Ref<TInstance>;
};

export type WithInnerRefProp<TProps, TInstance> = TProps & InnerRefProp<TInstance>;

export type ApplyFunction = () => string | undefined;

export default class GlitzBase<TProps> extends React.Component<ExternalProps<TProps>> {
  protected renderer: (apply: ApplyFunction) => React.ReactElement<any>;
  protected createApplier: (context: Context) => ApplyFunction;
  constructor(inner: string | React.ComponentType<TProps>, statics: StyleArray, props: ExternalProps<TProps>) {
    super(props);

    let lastContext: Context | undefined;
    let lastApplier: ApplyFunction | undefined;

    this.createApplier = (context: Context): ApplyFunction => {
      if (lastContext === context && lastApplier) {
        return lastApplier;
      }

      let cache: string | null = null;

      lastContext = context;

      return (lastApplier = () => {
        const styles: StyleOrStyleArray = compose();

        if (!styles) {
          return;
        }

        const isPure = styles === statics;

        if (isPure && cache) {
          return cache;
        }

        const classNames = context.glitz.injectStyle(styles);

        cache = isPure ? classNames : null;

        return classNames;
      });
    };

    const compose = (additionals?: StyleOrStyleArray): StyleOrStyleArray => {
      const dynamics: StyleOrStyleArray | undefined = this.props.css;

      if (!dynamics && !additionals) {
        return statics;
      }

      const styles = ([] as StyleArray).concat(additionals || [], statics, dynamics || []);

      return styles;
    };

    this.renderer =
      typeof inner === 'string' || (inner as any)[APPLY_CLASS_NAME_IDENTIFIER]
        ? (apply: ApplyFunction) => {
            const className = (this.props as any).className ? (this.props as any).className + ' ' + apply() : apply();
            const passProps = passingProps<TProps & StyledElementProps>({ className }, this.props);
            return React.createElement(inner, passProps);
          }
        : (apply: ApplyFunction) => {
            const passProps = passingProps<TProps & StyledProps>({ apply, compose }, this.props);
            return React.createElement(inner, passProps);
          };
  }
  public render() {
    return React.createElement(Consumer, null, (context: Context) => {
      if (process.env.NODE_ENV !== 'production') {
        if (!(context && context.glitz && context.glitz)) {
          throw new Error(
            "The `<GlitzProvider>` doesn't seem to be used correctly because the core instance couldn't be found",
          );
        }
      }

      return this.renderer(this.createApplier(context));
    });
  }
}

export function isStyledComponent<TProps>(
  inner: string | StyledComponent<TProps> | React.ComponentType<TProps>,
): inner is StyledComponent<TProps> {
  return typeof inner === 'function' && inner.prototype instanceof GlitzBase;
}

function passingProps<T>(destination: any, props: any): T {
  for (let name in props) {
    const value = props[name];
    if (name !== 'css') {
      if (name === 'innerRef') {
        name = 'ref';
      }
      // Don't override preexisting props
      destination[name] = destination[name] || value;
    }
  }
  return destination;
}
