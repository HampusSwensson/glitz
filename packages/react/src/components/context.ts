import { GlitzClient, GlitzServer } from '@glitz/core';
import * as React from 'react';

type Default = {
  glitz: null;
};

export type Context = {
  glitz: GlitzClient | GlitzServer;
};

export const { Provider, Consumer } = React.createContext<Default | Context>({ glitz: null });
