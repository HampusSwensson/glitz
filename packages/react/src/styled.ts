import { createStyled } from './styled/custom';
import { assignPredefined } from './styled/predefined';

export type { StyledComponent, StyledComponentWithRef } from './styled/create';
export type { StyledDecorator } from './styled/decorator';
export type { StyledElement, StyledElementProps } from './styled/predefined';
export type { StyledElementLike } from './styled/apply-class-name';
export type { StyledProps, StyledForwardStyle } from './styled/forward-style';
export { default as useStyle } from './styled/use-glitz';
export { default as useTheme } from './styled/use-theme';
export { applyClassName } from './styled/apply-class-name';
export { forwardStyle } from './styled/forward-style';

export const styled = assignPredefined(createStyled);
