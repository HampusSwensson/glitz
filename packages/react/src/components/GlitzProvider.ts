import type { GlitzClient, GlitzServer } from '@glitz/core';
import { PropsWithChildren, createElement } from 'react';
import { GlitzContext, StreamContext } from './context';

export type PropType = {
  glitz: GlitzClient | GlitzServer;
  stream?: boolean;
};

export function GlitzProvider(props: PropsWithChildren<PropType>) {
  return createElement(
    GlitzContext.Provider,
    { value: props.glitz },
    createElement(StreamContext.Provider, { value: !!props.stream }, props.children),
  );
}
