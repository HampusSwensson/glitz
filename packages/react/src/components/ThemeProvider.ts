import type { Theme } from '@glitz/core';
import { PropsWithChildren, createElement } from 'react';
import { ThemeContext } from './context';

type PropType = {
  theme: Theme;
};

export function ThemeProvider(props: PropsWithChildren<PropType>) {
  return createElement(ThemeContext.Provider, { value: props.theme }, props.children);
}
