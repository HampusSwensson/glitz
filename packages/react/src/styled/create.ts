import { Style } from '@glitz/type';
import * as React from 'react';
import { Context } from '../GlitzProvider';

export interface StyledComponent<TProps> extends React.ComponentClass<TProps & CSSProp & InnerRefProp> {}

export type StyledProps = {
  apply: () => string | undefined;
  compose: (style?: Style) => Style | Style[];
};

export type StyledElementProps = {
  className?: string;
};

export type CSSProp = {
  // `any[]` details type error details on `Style`, should be `Style[]` when conditional types are released
  css?: Style | any[];
};

export type InnerRefProp = {
  innerRef?: React.Ref<any>;
};

type InnerType<TProps> = React.ComponentType<TProps & StyledProps> | StyledComponent<TProps> | string;

const ASSIGN_METHOD = '__GLITZ_ASSIGN';

export default function create<TProps>(
  inner: React.ComponentType<TProps & StyledProps> | StyledComponent<TProps> | string,
  originalStaticStyle: Style[],
): StyledComponent<TProps> {
  if (isStyledComponent(inner)) {
    // @ts-ignore
    return inner[ASSIGN_METHOD](originalStaticStyle);
  }

  class GlitzStyled extends React.Component<TProps & CSSProp & InnerRefProp> {
    public static contextTypes = {
      glitz: () => null, // Just pass the damn thing
    };
    public static displayName: string;
    public static [ASSIGN_METHOD](assigningStyle?: Style) {
      return create(inner, assigningStyle ? originalStaticStyle.concat(assigningStyle) : originalStaticStyle);
    }
    constructor(props: TProps, context: Context) {
      super(props, context);

      if (process.env.NODE_ENV !== 'production') {
        if (!(context && context.glitz && context.glitz.glitz)) {
          throw new Error(
            "The `<GlitzProvider>` doesn't seem to be used correctly because the core instance couldn't be found",
          );
        }
      }

      const staticStyle: Style | Style[] = context.glitz.options.enableDeepComposition
        ? originalStaticStyle
        : flatten(originalStaticStyle);

      let cache: string | null = null;

      const apply = (): string | undefined => {
        const styles: Style | Style[] = compose();

        if (!styles) {
          return;
        }

        const isPure = styles === staticStyle;

        if (isPure && cache) {
          return cache;
        }

        const classNames = context.glitz.glitz.injectStyle(styles);

        cache = isPure ? classNames : null;

        return classNames;
      };

      const compose = (additionalStyle?: Style): Style | Style[] => {
        const dynamicStyle: Style | Style[] | undefined = this.props.css;

        if (!dynamicStyle && !additionalStyle) {
          return staticStyle;
        }

        const styles = ([] as Style[]).concat(additionalStyle || [], staticStyle, dynamicStyle || []);

        if (context.glitz.options.enableDeepComposition) {
          return styles;
        } else {
          return flatten(styles);
        }
      };

      this.render = isCustomComponent(inner)
        ? () => {
            const passProps: TProps & StyledProps = passingProps(this.props);

            passProps.apply = apply;
            passProps.compose = compose;

            return React.createElement(inner, passProps);
          }
        : () => {
            const passProps: TProps & StyledElementProps = passingProps(this.props);

            passProps.className = passProps.className ? passProps.className + ' ' + apply() : apply();

            return React.createElement(inner, passProps);
          };
    }
  }

  if (process.env.NODE_ENV !== 'production') {
    GlitzStyled.displayName = `Styled(${
      typeof inner === 'string' ? inner : inner.displayName || inner.name || 'Unknown'
    })`;
  }

  return GlitzStyled;
}

function flatten(styles: Style[]) {
  const result: Style = {};
  for (const chunk of styles) {
    let property: keyof Style;
    for (property in chunk) {
      result[property] = chunk[property];
    }
  }

  return result;
}

function passingProps(props: any) {
  const newProps = { ...props };

  delete newProps.css;

  if (newProps.innerRef) {
    newProps.ref = newProps.innerRef;
    delete newProps.innerRef;
  }

  return newProps;
}

function isStyledComponent<TProps>(inner: InnerType<TProps>): inner is StyledComponent<TProps> {
  return typeof inner === 'function' && ASSIGN_METHOD in inner;
}

function isCustomComponent<TProps>(inner: InnerType<TProps>): inner is React.ComponentType<TProps & StyledProps> {
  return typeof inner === 'function';
}
