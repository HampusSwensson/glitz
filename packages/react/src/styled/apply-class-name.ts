import { FunctionComponent, ComponentClass, ClassType, Component, ComponentState, ComponentType } from 'react';
import { SECRET_GLITZ_PROPERTY, ELEMENT_LIKE_TYPE } from './constants';
import { WithRefProp } from './create';
import { StyledType } from './predefined';
import { StyledElementProps } from './types';

export interface StyledElementLike<TComponent> extends StyledType {
  [SECRET_GLITZ_PROPERTY]: typeof ELEMENT_LIKE_TYPE;
  value: TComponent;
}

export function applyClassName<TProps extends StyledElementProps = StyledElementProps>(
  component: FunctionComponent<TProps>,
): StyledElementLike<FunctionComponent<TProps>>;

export function applyClassName<TProps extends StyledElementProps, TComponent extends ComponentClass<TProps>>(
  component: ClassType<TProps, Component<TProps, ComponentState>, TComponent>,
): StyledElementLike<TComponent>;

export function applyClassName<TProps extends StyledElementProps>(
  component: ComponentType<TProps>,
): StyledElementLike<ComponentType<TProps>> {
  return {
    [SECRET_GLITZ_PROPERTY]: ELEMENT_LIKE_TYPE,
    value: component,
  };
}

export function isElementLikeType<TProps extends StyledElementProps, TInstance>(
  type: any,
): type is StyledElementLike<ComponentType<WithRefProp<TProps, TInstance>>> {
  return SECRET_GLITZ_PROPERTY in type && type[SECRET_GLITZ_PROPERTY] === ELEMENT_LIKE_TYPE;
}
