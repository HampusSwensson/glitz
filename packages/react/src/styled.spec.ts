// tslint:disable max-classes-per-file

import { GlitzClient } from '@glitz/core';
import { mount } from 'enzyme';
import * as React from 'react';
import { GlitzProvider } from './';
import useGlitz from './styled/use-glitz';
import { styled } from './styled';
import { applyClassName } from './styled/apply-class-name';
import { StyledElementProps } from './styled/types';
import { ThemeProvider } from './components/ThemeProvider';
import { WithRefProp } from './styled/create';

describe('react styled', () => {
  const warn = console.warn;
  afterEach(() => {
    console.warn = warn;
  });

  it('returns class names using style hook', () => {
    function Component() {
      const className = useGlitz({ color: 'red' });
      expect(className).toBe('a');
      return null;
    }
    mountWithGlitz(React.createElement(Component));
  });
  it("renders styled component with no class names when there's no style", () => {
    const StyledComponent = styled(() =>
      React.createElement(styled.Div, {
        ref: (el: HTMLDivElement) => expect(el.getAttribute('class')).toBeNull(),
      }),
    );
    mountWithGlitz(React.createElement(StyledComponent));
  });
  it('renders styled component with class names with static style', () => {
    const StyledComponent = styled(
      () =>
        React.createElement(styled.Div, {
          ref: (el: HTMLDivElement) => expect(el.getAttribute('class')).toBe('a'),
        }),
      { color: 'red' },
    );
    mountWithGlitz(React.createElement(StyledComponent));
  });
  it('renders styled component with class names with css prop style', () => {
    const StyledComponent = styled(() =>
      React.createElement(styled.Div, {
        ref: (el: HTMLDivElement) => expect(el.getAttribute('class')).toBe('a'),
      }),
    );
    mountWithGlitz(React.createElement(StyledComponent, { css: { color: 'red' } }));
  });
  it('renders styled component with no class names on child element', () => {
    const StyledComponent = styled(() =>
      React.createElement(
        styled.Div,
        {
          ref: (el: HTMLDivElement) => expect(el.getAttribute('class')).toBe('a'),
        },
        React.createElement(styled.Div, {
          ref: (el: HTMLDivElement) => expect(el.getAttribute('class')).toBeNull(),
        }),
      ),
    );
    mountWithGlitz(React.createElement(StyledComponent, { css: { color: 'red' } }));
  });
  it('re-renders styled component with memoized class names', async () => {
    let updates = 0;
    let renders = 0;
    const StyledComponent = styled(
      React.memo(() => {
        renders++;
        return React.createElement(styled.Div, {
          ref: (el: HTMLDivElement) => expect(el.getAttribute('class')).toBe('a'),
        });
      }),
      { color: 'red' },
    );
    mountWithGlitz(
      React.createElement(
        class extends React.Component {
          public componentDidMount() {
            expect(renders).toBe(1);
            this.forceUpdate();
          }
          public render() {
            updates++;
            return React.createElement(StyledComponent);
          }
        },
      ),
    );
    expect(updates).toBe(2);
    expect(renders).toBe(1);
  });
  it('renders styled component with forwardRef', () => {
    const StyledComponent = styled(
      React.forwardRef(({}: {}, ref: React.Ref<HTMLDivElement>) =>
        React.createElement(styled.Div, {
          ref,
        }),
      ),
      { color: 'red' },
    );
    mountWithGlitz(
      React.createElement(StyledComponent, { ref: (el: HTMLDivElement) => expect(el.className).toBe('a') }),
    );
  });
  it('creates decorated styled component', () => {
    const styleDecoratorA = styled({ color: 'red' });
    const styleDecoratorB = styleDecoratorA({ backgroundColor: 'green' });
    const StyledComponentA = styleDecoratorB(() => {
      return React.createElement(styled.Div, {
        ref: (el: HTMLDivElement) => expect(el.className).toBe('a b'),
      });
    });
    const StyledComponentB = styleDecoratorB(
      () => {
        return React.createElement(styled.Div, {
          ref: (el: HTMLDivElement) => expect(el.className).toBe('b c'),
        });
      },
      { color: 'green' },
    );
    mountWithGlitz(
      React.createElement(
        React.Fragment,
        null,
        React.createElement(StyledComponentA),
        React.createElement(StyledComponentB),
      ),
    );
  });
  it('creates predefined styled component', () => {
    const StyledComponent = styled.div({ color: 'red' });
    mountWithGlitz(
      React.createElement(StyledComponent, {
        ref: (el: HTMLDivElement) => expect(el.className).toBe('a'),
      }),
    );
  });
  it('decorates style', () => {
    const styleDecoratorA = styled({ color: 'red' });
    const styleDecoratorB = styleDecoratorA({ backgroundColor: 'green' });
    const styleDecoratorC = styleDecoratorB(styled({ borderLeftColor: 'blue' })());
    expect(styleDecoratorC()).toEqual([{ color: 'red' }, { backgroundColor: 'green' }, { borderLeftColor: 'blue' }]);
    const StyledComponentA = styled(() => {
      return React.createElement(styled.Div, {
        css: styleDecoratorC,
        ref: (el: HTMLDivElement) => {
          expect(el.className).toBe('a b c');
          const declaration = getComputedStyle(el);
          expect(declaration.getPropertyValue('color')).toBe('red');
          expect(declaration.getPropertyValue('background-color')).toBe('green');
          expect(declaration.getPropertyValue('border-left-color')).toBe('blue');
        },
      });
    });
    mountWithGlitz(React.createElement(StyledComponentA));

    const StyledComponentB = styled(() => {
      return React.createElement(styled.Div, {
        css: styleDecoratorB({ color: 'blue' }),
        ref: (el: HTMLDivElement) => {
          expect(el.className).toBe('a b');
          const declaration = getComputedStyle(el);
          expect(declaration.getPropertyValue('background-color')).toBe('green');
          expect(declaration.getPropertyValue('color')).toBe('blue');
        },
      });
    });
    mountWithGlitz(React.createElement(StyledComponentB));

    const treeA = mountWithGlitz(React.createElement(styled.Div, { css: styleDecoratorB }));
    const declarationA = getComputedStyle(treeA.getDOMNode());
    expect(declarationA.getPropertyValue('color')).toBe('red');
    expect(declarationA.getPropertyValue('background-color')).toBe('green');

    const treeB = mountWithGlitz(React.createElement(styled.Div, { css: styleDecoratorB({ color: 'blue' }) }));
    const declarationB = getComputedStyle(treeB.getDOMNode());
    expect(declarationB.getPropertyValue('background-color')).toBe('green');
    expect(declarationB.getPropertyValue('color')).toBe('blue');
  });
  it('applies class names to element-like components', () => {
    const StyledComponent = styled(
      applyClassName(props => {
        expect(props.className).toBe('a');
        return React.createElement('div');
      }),
      { color: 'red' },
    );
    mountWithGlitz(React.createElement(StyledComponent));
  });
  it('renders predefined styled component', () => {
    const Component = () =>
      React.createElement(styled.Div, {
        css: { color: 'red' },
        ref: (el: HTMLDivElement) => expect(el.className).toBe('a'),
      });
    mountWithGlitz(React.createElement(Component));
  });
  it('composes style', () => {
    const StyledComponent = styled.div({ color: 'red' });
    const treeA = mountWithGlitz(
      React.createElement(StyledComponent, {
        css: { color: 'green', backgroundColor: 'red' },
      }),
    );
    const declarationA = getComputedStyle(treeA.getDOMNode());
    expect(declarationA.getPropertyValue('color')).toBe('green');
    expect(declarationA.getPropertyValue('background-color')).toBe('red');

    const ComposedComponentA = styled(
      () =>
        React.createElement(StyledComponent, {
          css: { color: 'green', backgroundColor: 'red' },
        }),
      { color: 'blue' },
    );
    const treeB = mountWithGlitz(React.createElement(ComposedComponentA));
    const declarationB = getComputedStyle(treeB.getDOMNode());
    expect(declarationB.getPropertyValue('color')).toBe('blue');
    expect(declarationB.getPropertyValue('background-color')).toBe('red');

    const ComposedComponentB = styled(
      () =>
        React.createElement(StyledComponent, {
          css: { color: 'green', backgroundColor: 'red' },
        }),
      { color: 'blue' },
    );
    const treeC = mountWithGlitz(
      React.createElement(ComposedComponentB, {
        css: { color: 'white' },
      }),
    );
    const declarationC = getComputedStyle(treeC.getDOMNode());
    expect(declarationC.getPropertyValue('color')).toBe('white');
    expect(declarationC.getPropertyValue('background-color')).toBe('red');

    const ComposedComponentC = styled(
      () =>
        React.createElement(styled.Div, {
          css: { color: 'green', backgroundColor: 'red' },
        }),
      { color: 'red' },
    );
    const treeD = mountWithGlitz(React.createElement(ComposedComponentC));
    const declarationD = getComputedStyle(treeD.getDOMNode());
    expect(declarationD.getPropertyValue('color')).toBe('red');
    expect(declarationD.getPropertyValue('background-color')).toBe('red');
  });
  it('caches pure style', () => {
    let parses = 0;
    let renders = 0;
    const StyledComponent = styled(() => {
      const [, setState] = React.useState(0);
      React.useEffect(() => setState(1), []);
      renders++;
      return React.createElement(styled.Div, {
        css: {
          get color() {
            parses++;
            return 'red';
          },
        },
      });
    });
    mountWithGlitz(React.createElement(StyledComponent));
    expect(parses).toBe(2);
    expect(renders).toBe(2);

    parses = 0;
    renders = 0;
    const StyledComponentPure = styled(() => {
      const [, setState] = React.useState(0);
      React.useEffect(() => setState(1), []);
      renders++;
      return React.createElement(styled.Div, { css: cssA1 });
    });
    const cssA1 = {
      get color() {
        parses++;
        return 'red';
      },
    };
    const cssA2 = {};
    mountWithGlitz(React.createElement(StyledComponentPure, { css: cssA2 }));
    expect(parses).toBe(1);
    expect(renders).toBe(2);

    parses = 0;
    renders = 0;
    const StyledComponentMemo = styled(() => {
      const [, setState] = React.useState(0);
      renders++;
      const cssB = React.useMemo(
        () => ({
          get color() {
            parses++;
            return 'red';
          },
        }),
        [],
      );
      React.useEffect(() => setState(1), []);
      return React.createElement(styled.Div, { css: cssB });
    });
    mountWithGlitz(React.createElement(StyledComponentMemo));
    expect(parses).toBe(1);
    expect(renders).toBe(2);
  });
  it('renders correctly with new instance', () => {
    const StyledComponentA = styled.div({
      color: 'red',
    });
    const StyledComponentB = styled.div({
      color: 'green',
    });
    mount(
      React.createElement(
        'div',
        {},
        React.createElement(
          GlitzProvider,
          {
            glitz: new GlitzClient(),
          },
          React.createElement(StyledComponentA, {
            ref: (el: HTMLDivElement) => expect(el.className).toBe('a'),
          }),
          React.createElement(StyledComponentB, {
            ref: (el: HTMLDivElement) => expect(el.className).toBe('b'),
          }),
        ),
        React.createElement(
          GlitzProvider,
          {
            glitz: new GlitzClient(),
          },
          React.createElement(StyledComponentB, {
            ref: (el: HTMLDivElement) => expect(el.className).toBe('a'),
          }),
          React.createElement(StyledComponentA, {
            ref: (el: HTMLDivElement) => expect(el.className).toBe('b'),
          }),
        ),
      ),
    );
  });
  it('passes ref prop', () => {
    class Spy extends React.Component<StyledElementProps> {
      public render() {
        return React.createElement('div');
      }
    }
    const StyledComponent = styled(Spy);
    const StyledElementLike = styled(applyClassName(Spy));
    mountWithGlitz(
      React.createElement(
        React.Fragment,
        null,
        React.createElement(StyledComponent, {
          ref: (el: Spy) => {
            expect(el).toBeInstanceOf(Spy);
          },
        }),
        React.createElement(StyledElementLike, {
          ref: (el: Spy) => {
            expect(el).toBeInstanceOf(Spy);
          },
        }),
        React.createElement(styled.Div, {
          ref: (el: HTMLDivElement) => {
            expect(el).toBeInstanceOf(HTMLDivElement);
          },
        }),
      ),
    );
  });
  it('passes className prop', () => {
    mountWithGlitz(
      React.createElement(styled.Div, {
        className: 'some-class',
        css: { color: 'red' },
        ref: (el: HTMLDivElement) => {
          expect(el.className).toBe('some-class a');
        },
      }),
    );

    const StyledElementLike = styled(
      applyClassName(props => {
        expect(props.className).toBe('some-class a');
        return React.createElement('div');
      }),
      { color: 'red' },
    );
    mountWithGlitz(
      React.createElement(StyledElementLike, {
        className: 'some-class',
      }),
    );

    const StyledElement = styled.div({ color: 'red' });
    mountWithGlitz(
      React.createElement(StyledElement, {
        className: 'some-class',
        ref: (el: HTMLDivElement) => {
          expect(el.className).toBe('some-class a');
        },
      }),
    );

    class ClassNameUpdates extends React.Component {
      public state = {
        className: 'some-class',
      };
      public render() {
        return React.createElement(StyledElement, {
          className: this.state.className,
        });
      }
    }

    const tree = mountWithGlitz(React.createElement(ClassNameUpdates));

    expect(tree.getDOMNode().className).toBe('some-class a');

    tree.setState({ className: 'another-class' });

    expect(tree.getDOMNode().className).toBe('another-class a');
  });
  it('updates when theme changes', () => {
    class Theming extends React.Component {
      public state = {
        color: 'red',
      };
      public render() {
        return React.createElement(ThemeProvider, { theme: this.state }, this.props.children);
      }
    }

    const StyledComponent = styled(styled.Div, {
      color(theme: any) {
        return theme.color;
      },
    });

    const tree = mountWithGlitz(React.createElement(Theming, null, React.createElement(StyledComponent)));

    expect(tree.getDOMNode().className).toBe('a');

    tree.setState({ color: 'green' });

    expect(tree.getDOMNode().className).toBe('b');
  });
  it('styles with HOC in between', () => {
    function hoc<TProps, TInstance>(
      Component: React.ComponentType<WithRefProp<TProps, TInstance>> | React.ComponentType<TProps>,
    ) {
      return React.forwardRef<TInstance, TProps>((props, ref) =>
        React.createElement<any>(Component, { ...props, ref }),
      );
    }
    const ConnectedStyledComponentA = hoc(
      styled(
        ({}: {}) =>
          React.createElement(styled.Div, {
            ref: (el: HTMLDivElement) => expect(el.className).toBe('a b'),
          }),
        { color: 'red' },
      ),
    );
    mountWithGlitz(React.createElement(ConnectedStyledComponentA, { css: { backgroundColor: 'green' } }));

    const ConnectedStyledComponentB = hoc(styled.div({ color: 'red' }));
    mountWithGlitz(
      React.createElement(ConnectedStyledComponentB, {
        css: { backgroundColor: 'green' },
        ref: (el: HTMLDivElement) => expect(el.className).toBe('a b'),
      }),
    );
  });
  it('warns of multiple re-renders with invalidated cache', async () => {
    const createAnimationSimulator = (
      count: number,
      update: (callback?: (() => void) | undefined) => void,
      resolve: () => void,
    ) => () => {
      let updates = 0;
      const frame = () =>
        requestAnimationFrame(() => {
          if (updates < count) {
            updates++;
            update(frame);
          } else {
            requestAnimationFrame(() => resolve());
          }
        });
      frame();
    };

    let renders = 0;
    const logger = (console.warn = jest.fn());
    await new Promise(resolve => {
      const css = {
        color: 'red',
      };
      const CacheValidated = class extends React.Component {
        public componentDidMount = createAnimationSimulator(5, callback => this.forceUpdate(callback), resolve);
        public render() {
          renders++;
          return React.createElement(styled.Div, { css });
        }
      };
      mountWithGlitz(React.createElement(CacheValidated));
    });
    expect(renders).toBe(6);
    expect(logger).toHaveBeenCalledTimes(0);

    renders = 0;
    await new Promise(resolve => {
      const CacheInvalidated = class extends React.Component {
        public componentDidMount = createAnimationSimulator(4, callback => this.forceUpdate(callback), resolve);
        public render() {
          renders++;
          return React.createElement(styled.Div, { css: { color: 'red' } });
        }
      };
      mountWithGlitz(React.createElement(CacheInvalidated));
    });
    expect(renders).toBe(5);
    expect(logger).toHaveBeenCalledTimes(0);

    renders = 0;
    await new Promise(resolve => {
      const CacheInvalidated = class extends React.Component {
        public componentDidMount = createAnimationSimulator(5, callback => this.forceUpdate(callback), resolve);
        public render() {
          renders++;
          return React.createElement(styled.Div, { css: { color: 'red' } });
        }
      };
      mountWithGlitz(React.createElement(CacheInvalidated));
    });
    expect(renders).toBe(6);
    expect(logger).toHaveBeenCalledTimes(1);
  });
  it('inherits default props when composing', () => {
    const StyledComponent = styled(styled.Button);
    StyledComponent.defaultProps = { type: 'button' };

    const ComposedComponent = styled(StyledComponent);
    expect(ComposedComponent.defaultProps).toEqual({ type: 'button' });
  });
  it('inherits display name when composing', () => {
    const StyledComponent = styled(styled.Button);
    expect(StyledComponent.displayName).toBe('Styled(button)');

    StyledComponent.displayName = 'Test';

    const ComposedComponent = styled(StyledComponent);
    expect(ComposedComponent.displayName).toBe('Test');
  });
  it('does not trigger update on child elements when style changes', () => {
    let renders = 0;
    const Child = styled(
      React.memo(() => {
        renders++;
        return React.createElement(styled.Div);
      }),
    );

    const StyledComponent = styled(() => React.createElement(styled.Div, null, React.createElement(Child)));

    const css = { color: 'red' };

    const tree = mountWithGlitz(React.createElement(StyledComponent, { css }));
    expect(renders).toBe(1);

    tree.setProps({ css });
    expect(renders).toBe(1);

    tree.setProps({ css: { color: 'red' } });
    expect(renders).toBe(1);

    tree.setProps({ css: { color: 'green' } });
    expect(renders).toBe(1);
  });
  it('creates an empty decorator', () => {
    let decorator = styled();

    decorator = decorator({ color: 'red' });
    decorator = decorator({ backgroundColor: 'green' });
    expect(decorator()).toEqual([{ color: 'red' }, { backgroundColor: 'green' }]);
  });
});

function mountWithGlitz(node: React.ReactElement) {
  return mount(node, {
    wrappingComponent: GlitzProvider,
    wrappingComponentProps: {
      glitz: new GlitzClient(),
    },
  });
}
