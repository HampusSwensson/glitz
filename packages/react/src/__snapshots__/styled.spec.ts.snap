// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react styled creates predefined styled component 1`] = `
<div
  className="a"
/>
`;

exports[`react styled renders predefined styled component 1`] = `
<div
  className="a"
/>
`;
